### ISSUE TITLE
Update Function component

### ISSUE THREAD

**Issue by eahefnawy** [serverless/components#440]
Update the Function component to take restApi id, path & method and ship it!

**Comment by eahefnawy** [comment]
Thinking about this a bit deeper, one worry I have regarding reusing existing restApis is that there could be a race condition involved when a user is using multiple Function instances that are running in parallel.

Because of how APIG handles deployment, whenever a Function component adds endpoints/methods, they have to explicitly `createDeployment`. So I'm worried that running this operation in parallel could eventually result in a deployment that is missing some endpoints.

The way this is meant to be done in the AWS world is that we just run `createDeployment` once all endpoints/methods has been added. I hope we won't need an `apig-deployment` component to get this working reliably 😒

**Comment by eahefnawy** [comment]
Another challenge. With the following setup:

**Function A:**
```yaml
component: "@serverless/function"
inputs:
    path: users/create
    method: post
```

**Function B:**
```yaml
component: "@serverless/function"
inputs:
    path: users/list
    method: get
```

Which function owns the `users` path part? In other words, if you remove function A, should it also remove the `users` path part resource which function B depends on? Cause that would def break function B.

A possible solution is to have each function own only the last path part. So in this example, function A would own the `create` path part, and function B would own the `list` path part. But then who owns the `users` path part? This is all due to how AWS APIG is designed. Each path part needs to be created individually.

@ac360 in #432 you mentioned that each function owns an "endpoint", the issue is that a single endpoint could (in most cases) contain multiple path parts, which are different resources on aws.

**Comment by eahefnawy** [comment]
Another possible solution is to have a single function own the entire path. So in this example, function A would own both `users/create` and `users/X`. Which means:

1. Function B would throw an error
2. Function A could now accept multiple endpoints/methods, which should each point to a separate function, effectively making the function component a microservice compoent with multiple functions.

**Comment by pmuens** [comment]
> Because of how APIG handles deployment, whenever a Function component adds endpoints/methods, they have to explicitly createDeployment. So I'm worried that running this operation in parallel could eventually result in a deployment that is missing some endpoints.

Hmm. I see the potential problems with the parallel deployments. Looking into the AWS SDK docs there's an [`updateDeployment`](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/APIGateway.html#updateDeployment-property) method we could use. Not sure how this behaves if you have multiple deployments so we might need to queue this somehow 🤔

> Which function owns the users path part? In other words, if you remove function A, should it also remove the users path part resource which function B depends on? Cause that would def break function B.

I'd propose that every functions own the whole thing, meaning that in your example if Function A is removed there's still Function B owning the path part. Once Function B is removed and there are no more paths with `users` it should remove the path as well.

**Comment by eahefnawy** [comment]
Ok I think I found a better solution:

### on deploy
when deploying a function with a certain endpoint/method, it checks if that **exact** endpoint/method pair exists or not, if it does, it'd throw an error, otherwise deployment is successful.

### on remove
when removing a function with a certain endpoint/method, it checks each path part separately and see if it has any dependents (e.g. a child path, or method not controlled by this function). If dependents exist, then we'll only remove the function and the correspond http method, but not the path (even though it technically created the entire path). If no dependencies found, it'll remove everything it created.

## Example

  1. deploy function A `post /users/create`
  2. deploy function B `get /users`
  3. removing function A would only remove the `create` part but leave the `users` part because it's shared with another function, even though it technically owns the `users` part too.
  4. removing function B would remove everything left.


This basically removes the concept of "endpoint ownership" that we've talked about, and rather convert it into "endpoint sharing" with some intelligent code that figures out what should be deployed/removed based on dependencies and the state on the provider.

This is gonna be pretty tough to implement though...

**Comment by eahefnawy** [comment]
> I'd propose that every functions own the whole thing, meaning that in your example if Function A is removed there's still Function B owning the path part. Once Function B is removed and there are no more paths with users it should remove the path as well.

@pmuens but doesn't function A technically own the `users` path part too and it should be removed along with it?

**Comment by ac360** [comment]
A couple of thoughts:

API Gateway Deployments seem like they will happen infrequently.  If the path/method isn't changed, no API Gateway createDeployment is necessary.  Unless we are going to bump the Lambda version on every deploy.  I'm not sure this is necessary though.

**Comment by ac360** [comment]
> deploy function A post /users/create
deploy function B get /users
removing function A would only remove the create part but leave the users part because it's shared with another function, even though it technically owns the users part too.
removing function B would remove everything left.
This basically removes the concept of "endpoint ownership" that we've talked about, and rather convert it into "endpoint sharing" with some intelligent code that figures out what should be deployed/removed based on dependencies and the state on the provider.

> This is gonna be pretty tough to implement though...

Hm, I think we just need to do 4 other API requests to see if the other methods are assigned to anything, right?

**Comment by eahefnawy** [comment]
@ac360 true, my worry is the initial deployment of all the Function components defined in a single serverless.yml file. We just gotta make sure the final 'createDeployment` call won't overwrite previous ones

**Comment by ac360** [comment]
A few notes from our conversation:

Component - REST API

- Should create new or work with existing REST APIs
- Should optionally accept a Swagger definition.
- Should declare all routes/methods/global settings
- Should inherit Function Components as dependencies

Component - Users Management Service

- Exports Functions and Outputs

Component - Passwordless

- How can developers deploy this quickly to reach their desired outcome?
- How can this be used within a larger team context and a shared REST API?

**Comment by eahefnawy** [comment]
@ac360 thinking about it a bit deeper, it seems what you're suggesting is what we already have 🤔 ... We have the `aws-api-gateway` component, which declares all the endpoints mapped to function ARNs. Those ARNs could be existing functions or functions defined in the same yam l file.

So with this setup, people could create functions/microservices independently, then the "gate keeper", that is apig component, could just map that to an endpoint and deploy it. Is that what we want?

If that's the case, then we just need to update APIG component to work with existing rest apis. If an API does exist, we're gonna keep track of the endpoints this component owns/creates so that only those endpoints are removed when you run `components remove`. However this means we can't remove existing rest apis with components.

Honestly this is a broader question. For existing resources, do we want to be able to manage & remove them completely, or just "undo" what was done when we hooked up with them?!
